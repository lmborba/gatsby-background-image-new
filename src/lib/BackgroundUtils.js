/**
 * Gets styles by a class name.
 *
 * @notice The className has to exactly match the CSS class
 * @param className string
 */
import { isBrowser, isString, stringToArray, toCamelCase } from './SimpleUtils';

/**
 * Gets styles rules by a class name.
 *
 * @notice The className has to exactly match the CSS class
 * @param className string
 */
export const getStyleRulesForClassName = className => {
  const styleSheets = isBrowser() ? window.document.styleSheets : [];
  for (let i = 0; i < styleSheets.length; i++) {
    let classes;
    try {
      classes =
        typeof styleSheets[i].rules !== 'undefined'
          ? styleSheets[i].rules
          : typeof styleSheets[i].cssRules !== 'undefined'
          ? styleSheets[i].cssRules
          : '';
    } catch (e) {}
    if (classes) {
      const foundClass = Array.prototype.slice
        .call(classes)
        .reduce((foundAcc, styleRule) => {
          if (styleRule.selectorText === className) {
            foundAcc.push(styleRule);
          }
          return foundAcc;
        }, []);
      if (foundClass.length) {
        return foundClass;
      }
    }
  }
  return [];
};

/**
 * Fixes non-enumerable style rules in Firefox.
 *
 * @param cssStyleRules CSSStyleRules   DOM-StyleRules-Object
 * @return {*}
 */
export const getStyleRules = cssStyleRules => {
  let styles = {};
  if (
    cssStyleRules.length > 0 &&
    typeof cssStyleRules[0].style !== 'undefined'
  ) {
    // Fallback for Browsers without constructor.name (IE11).
    const constructorName =
      cssStyleRules[0].style.constructor.name ||
      cssStyleRules[0].style.constructor.toString();

    switch (constructorName) {
      // For Firefox or IE11.
      case 'CSS2Properties':
      case '[object MSStyleCSSProperties]':
        Object.values(cssStyleRules[0].style).forEach(prop => {
          styles[toCamelCase(prop)] = cssStyleRules[0].style[prop];
        });
        break;
      case 'CSSStyleDeclaration':
        styles = cssStyleRules[0].style;
        break;
      default:
        console.error('Unknown style object prototype');
        break;
    }
  }
  return styles;
};

/**
 * Filters out Background Rules for a given class Name.
 *
 * @param className   string    The class to filter rules from
 * @return {{}}
 */
export const getBackgroundStylesForSingleClass = className => {
  if (className && isString(className)) {
    const cssStyleRules = getStyleRulesForClassName(`.${className}`);
    // const cssStyleRules = rulesForCssText(style)

    if (
      cssStyleRules?.length > 0 &&
      typeof cssStyleRules[0].style !== 'undefined'
    ) {
      // Filter out background(-*) rules that contain a definition.
      return Object.keys(getStyleRules(cssStyleRules))
        .filter(
          key =>
            key.indexOf('background') === 0 &&
            cssStyleRules[0].style[key] !== ''
        )
        .reduce((newData, key) => {
          newData[toCamelCase(key)] = cssStyleRules[0].style[key];
          return newData;
        }, {});
    }
  }
  return {};
};

/**
 * Uses the above to get all background(-*) rules from given class(es).
 *
 * @param className   string|array    className or array of classNames
 * @return {*}
 */
const getBackgroundStyles = className => {
  //if (isBrowser()) {
    const classes = stringToArray(className);
    if (classes instanceof Array) {
      const classObjects = [];
      classes.forEach(item =>
        classObjects.push(getBackgroundStylesForSingleClass(item))
      );
      return Object.assign(...classObjects);
    }
    return getBackgroundStylesForSingleClass(className);
  //}
  //return {};
};

export default getBackgroundStyles;
