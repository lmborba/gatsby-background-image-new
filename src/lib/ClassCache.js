const componentClassCache = Object.create({});
/**
 * Cache component classes as we never know if a Component wasn't already set.
 *
 * @param className   string  className given by props
 * @return {*|boolean}
 */
export const inComponentClassCache = className => {
  return componentClassCache[className] || false;
};

/**
 * Adds a component's classes to componentClassCache.
 *
 * @param className   string  className given by props
 */
export const activateCacheForComponentClass = className => {
  if (className) {
    componentClassCache[className] = true;
  }
};

/**
 * Resets the componentClassCache (especially important for reliable tests).
 */
export const resetComponentClassCache = () => {
  for (const className in componentClassCache)
    delete componentClassCache[className];
};
