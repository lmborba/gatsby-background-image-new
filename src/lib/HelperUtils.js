import filterInvalidDOMProps from 'filter-invalid-dom-props';
import { groupByMedia, hasArtDirectionSupport } from './MediaUtils';

/**
 * Strip BackgroundImage propTypes from remaining props to be passed to <Tag />
 *
 * @param props
 * @return {Object}
 */
export const stripRemainingProps = props => filterInvalidDOMProps(props);

/**
 * Preprocess art directed images.
 *
 * @param props
 * @return {Object}
 */
export const convertProps = props => {
  const convertedProps = { ...props };
  const { fixed, fluid } = convertedProps;

  // if (fluid && !hasImageArray(props)) {
  //   convertedProps.fluid = [].concat(fluid)
  // }
  //
  // if (fixed && !hasImageArray(props)) {
  //   convertedProps.fixed = [].concat(fixed)
  // }

  // convert fluid & fixed to arrays so we only have to work with arrays
  if (fluid && hasArtDirectionSupport(props, 'fluid')) {
    convertedProps.fluid = groupByMedia(convertedProps.fluid);
  }
  if (fixed && hasArtDirectionSupport(props, 'fixed')) {
    convertedProps.fixed = groupByMedia(convertedProps.fixed);
  }

  return convertedProps;
};
