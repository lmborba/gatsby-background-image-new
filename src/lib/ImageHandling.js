import { createDummyImageArray, getCurrentFromData } from './ImageUtils';
import { hasArtDirectionArray } from './MediaUtils';
import { combineArray, filteredJoin } from './SimpleUtils';

/**
 * Compares the old states to the new and changes image settings accordingly.
 *
 * @param image     string||array   Base data for one or multiple Images.
 * @param bgImage   string||array   Last background image(s).
 * @param imageRef  string||array   References to one or multiple Images.
 * @param state     object          Component state.
 * @return {{afterOpacity: number, bgColor: *, bgImage: *, nextImage: string}}
 */
// eslint-disable-next-line import/prefer-default-export
export const switchImageSettings = ({ image, bgImage, imageRef, state }) => {
  // Read currentSrc from imageRef (if exists).
  const currentSources = getCurrentFromData({
    data: imageRef,
    propName: `currentSrc`,
  });
  // Check if image is Array.
  const returnArray =
    Array.isArray(image) && !hasArtDirectionArray({ fluid: image });
  // Backup bgImage to lastImage.
  const lastImage = Array.isArray(bgImage) ? filteredJoin(bgImage) : bgImage;
  // Set the backgroundImage according to images available.
  let nextImage;
  let nextImageArray;
  // Signal to `createPseudoStyles()` when we have reached the final image,
  // which is important for transparent background-image(s).
  let finalImage = returnArray && state.seenBefore && !!currentSources;
  if (returnArray) {
    if (!currentSources) {
      // Check for tracedSVG first.
      nextImage = getCurrentFromData({
        data: image,
        propName: `tracedSVG`,
        returnArray,
      });
      // Now combine with base64 images.
      nextImage = combineArray(
        getCurrentFromData({
          data: image,
          propName: `base64`,
          returnArray,
        }),
        nextImage
      );
    }
    // Now add possible `rgba()` or similar CSS string props.
    nextImage = combineArray(
      getCurrentFromData({
        data: image,
        propName: `CSS_STRING`,
        addUrl: false,
        returnArray,
      }),
      nextImage
    );
    // Do we have at least one img loaded?
    if ((state.imgLoaded || !!currentSources) && state.isVisible) {
      if (currentSources) {
        nextImage = combineArray(
          getCurrentFromData({
            data: imageRef,
            propName: `currentSrc`,
            returnArray,
          }),
          nextImage
        );
        finalImage = true;
      } else {
        // No support for HTMLPictureElement or WebP present, get src.
        nextImage = combineArray(
          getCurrentFromData({
            data: imageRef,
            propName: `src`,
            returnArray,
          }),
          nextImage
        );
        finalImage = true;
      }
    }
    // First fill last images from bgImage...
    nextImage = combineArray(nextImage, bgImage);
    // ... then fill the rest of the background-images with a transparent dummy
    // pixel, lest the background-* properties can't target the correct image.
    const dummyArray = createDummyImageArray(image.length);
    // Now combine the two arrays and join them.
    nextImage = combineArray(nextImage, dummyArray);
    nextImageArray = nextImage;
    nextImage = filteredJoin(nextImage);
  } else {
    nextImage = ``;
    nextImage =
      getCurrentFromData({ data: image, propName: `tracedSVG` }) ||
      getCurrentFromData({ data: image, propName: `base64` });
    if (state.imgLoaded && state.isVisible) {
      nextImage = currentSources;
      finalImage = true;
    }
  }

  // Change opacity according to imageState.
  const afterOpacity = state.imageState % 2;

  if (
    !returnArray &&
    nextImage === `` &&
    state.imgLoaded &&
    state.isVisible &&
    imageRef &&
    !imageRef.currentSrc
  ) {
    // Should we still have no nextImage it might be because currentSrc is missing.
    nextImage = getCurrentFromData({
      data: imageRef,
      propName: `src`,
      checkLoaded: false,
    });
    finalImage = true;
  }
  // Fall back on lastImage (important for prop changes) if all else fails.
  if (!nextImage) nextImage = lastImage;

  const newImageSettings = {
    lastImage,
    nextImage,
    afterOpacity,
    finalImage,
  };
  // Add nextImageArray for bgImage to newImageSettings if exists.
  if (nextImageArray) newImageSettings.nextImageArray = nextImageArray;
  return newImageSettings;
};
