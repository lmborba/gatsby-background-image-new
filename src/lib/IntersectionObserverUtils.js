import { isBrowser } from "./SimpleUtils";

let io;
const listeners = new WeakMap();

/**
 * Executes each IntersectionObserver entries' callback.
 *
 * @param entries
 */
export const callbackIO = entries => {
  entries.forEach(entry => {
    if (listeners.has(entry.target)) {
      const callback = listeners.get(entry.target);
      // Edge doesn't currently support isIntersecting, so also test for an intersectionRatio > 0
      if (entry.isIntersecting || entry.intersectionRatio > 0) {
        io.unobserve(entry.target);
        listeners.delete(entry.target);
        callback();
      }
    }
  });
};

/**
 * Returns an IntersectionObserver instance if exists.
 *
 * @param rootMargin    string    The current rootMargin, defaults to 200px.
 * @return {IntersectionObserver|undefined}
 */
export const getIO = rootMargin => {
  if (
    typeof io === `undefined` &&
    isBrowser() &&
    window.IntersectionObserver
  ) {
    io = new window.IntersectionObserver(callbackIO, {
      rootMargin,
    });
  }

  return io;
};

/**
 * Registers IntersectionObserver callback on element.
 *
 * @param element     Element   The element to observe.
 * @param callback    function  Callback to call when intersecting.
 * @param rootMargin  string    The current rootMargin, defaults to 200px.
 * @return {Function}
 */
export const listenToIntersections = (
  element,
  callback,
  rootMargin = `200px`
) => {
  const observer = getIO(rootMargin);

  if (observer) {
    observer.observe(element);
    listeners.set(element, callback);
    return () => {
      observer.unobserve(element);
      listeners.delete(element);
    };
  }
  return () => {};
};
