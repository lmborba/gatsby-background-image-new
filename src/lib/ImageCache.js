import { convertProps } from './HelperUtils';
import { hasArtDirectionArray } from './MediaUtils';
import { getImageSrcKey, getSelectedImage, hasImageArray } from './ImageUtils';
import { isObject } from './SimpleUtils';

const imageCache = Object.create({});
/**
 * Cache if we've seen an image before so we don't bother with
 * lazy-loading & fading in on subsequent mounts.
 *
 * @param props
 * @param index
 * @param isLoop
 * @return {boolean}
 */
export const inImageCache = (props, index = 0, isLoop = false) => {
  const convertedProps = convertProps(props);
  const isImageStack =
    hasImageArray(convertedProps) && !hasArtDirectionArray(convertedProps);
  if (isImageStack && !isLoop) {
    return allInImageCache(props);
  }

  // Find src
  const src = isImageStack
    ? getSelectedImage(convertedProps, index)
    : getImageSrcKey(convertedProps);
  if (isObject(src)) {
    const objectSrc = getImageSrcKey({ fluid: src, fixed: src });
    return imageCache[objectSrc] || false;
  }
  return imageCache[src] || false;
};

/**
 * Processes an image stack with inImageCache.
 *
 * @param props  object    Component Props (with fluid or fixed as array).
 * @return {*|boolean}
 */
export const allInImageCache = props => {
  const convertedProps = convertProps(props);
  // Extract Image Array.
  const imageStack = convertedProps.fluid || convertedProps.fixed;
  // Only return true if every image is in cache.
  return imageStack.every((imageData, index) => {
    return inImageCache(convertedProps, index, true);
  });
};

/**
 * Adds an Image to imageCache.
 *
 * @param props
 * @param index
 * @param isLoop
 */
export const activateCacheForImage = (props, index = 0, isLoop = false) => {
  const convertedProps = convertProps(props);
  const isImageStack =
    hasImageArray(convertedProps) && !hasArtDirectionArray(convertedProps);
  if (isImageStack && !isLoop) {
    return activateCacheForMultipleImages(props);
  }
  // Find src
  const src = isImageStack
    ? getSelectedImage(convertedProps, index)
    : getImageSrcKey(convertedProps);
  if (src) {
    if (isObject(src)) {
      const objectSrc = getImageSrcKey({ fluid: src, fixed: src });
      imageCache[objectSrc] = true;
    } else {
      imageCache[src] = true;
    }
  }
};

/**
 * Activates the Cache for multiple Images.
 *
 * @param props
 */
export const activateCacheForMultipleImages = props => {
  const convertedProps = convertProps(props);
  // Extract Image Array.
  const imageStack = convertedProps.fluid || convertedProps.fixed;
  imageStack.forEach((imageData, index) =>
    activateCacheForImage(convertedProps, index, true)
  );
};

/**
 * Resets the image cache (especially important for reliable tests).
 */
export const resetImageCache = () => {
  for (const prop in imageCache) delete imageCache[prop];
};
