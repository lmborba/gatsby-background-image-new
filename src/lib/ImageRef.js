import { convertProps } from './HelperUtils';
import {
  getCurrentSrcData,
  getSelectedImage,
  hasImageArray,
  hasPictureElement,
  imageLoaded,
} from './ImageUtils';
import {
  createArtDirectionSources,
  createSourceElementForSrcSet,
  hasArtDirectionArray,
} from './MediaUtils';
import { isBrowser, isString } from './SimpleUtils';
import { inImageCache } from './ImageCache';

/**
 * Creates an image reference to be activated on critical or visibility.
 * @param props
 * @param onLoad
 * @param index
 * @param isLoop
 * @return {HTMLImageElement|null|Array}
 */
export const createPictureRef = (props, onLoad, index, isLoop = false) => {
  const convertedProps = convertProps(props);

  if (
    isBrowser() &&
    (typeof convertedProps.fluid !== `undefined` ||
      typeof convertedProps.fixed !== `undefined`)
  ) {
    const isImageStack =
      hasImageArray(convertedProps) && !hasArtDirectionArray(convertedProps);
    if (isImageStack && !isLoop) {
      return createMultiplePictureRefs(props, onLoad);
    }
    const img = new Image();

    img.onload = () => onLoad();

    if (!img.complete && typeof convertedProps.onLoad === `function`) {
      img.addEventListener('load', convertedProps.onLoad);
    }
    if (typeof convertedProps.onError === `function`) {
      img.addEventListener('error', convertedProps.onError);
    }
    if (convertedProps.crossOrigin) {
      img.crossOrigin = convertedProps.crossOrigin;
    }

    // Only directly activate the image if critical (preload).
    if ((convertedProps.critical || convertedProps.isVisible) && !isLoop) {
      return activatePictureRef(img, convertedProps, index, isLoop);
    }

    return img;
  }
  return null;
};

/**
 * Creates multiple image references. Internal function.
 *
 * @param props   object    Component Props (with fluid or fixed as array).
 * @param onLoad  function  Callback for load handling.
 */
export const createMultiplePictureRefs = (props, onLoad) => {
  const convertedProps = convertProps(props);

  // Extract Image Array.
  const imageStack = convertedProps.fluid || convertedProps.fixed;
  const imageRefs = imageStack.map((imageData, index) =>
    createPictureRef(convertedProps, onLoad, index, true)
  );
  // Only directly activate the image if critical (preload).
  if (convertedProps.critical || convertedProps.isVisible) {
    return activatePictureRef(imageRefs, convertedProps);
  }
  return imageRefs;
};

/**
 * Creates a picture element for the browser to decide which image to load.
 *
 * @param imageRef
 * @param props
 * @param selfRef
 * @param index
 * @param isLoop
 * @return {null|Array|*}
 */
export const activatePictureRef = (
  imageRef,
  props,
  selfRef = null,
  index = 0,
  isLoop = false
) => {
  const convertedProps = convertProps(props);
  if (
    isBrowser() &&
    (typeof convertedProps.fluid !== `undefined` ||
      typeof convertedProps.fixed !== `undefined`)
  ) {
    const isImageStack =
      hasImageArray(convertedProps) && !hasArtDirectionArray(convertedProps);
    if (isImageStack && !isLoop) {
      return activateMultiplePictureRefs(imageRef, props, selfRef);
    }
    // Clone body to get the correct sizes.
    // const bodyClone = document.body.cloneNode(true)
    const bodyClone = document.createElement('body');
    // Do we have an image stack or Art-direction array?
    // Then get its current or first(smallest) image respectively.
    const imageData = isImageStack
      ? getSelectedImage(convertedProps, index)
      : getCurrentSrcData(convertedProps);

    if (!imageData) {
      return null;
    }
    if (isString(imageData)) {
      return imageData;
    }

    if (selfRef) {
      // Set original component's style.
      imageRef.width = selfRef.offsetWidth;
      imageRef.height = selfRef.offsetHeight;
    }

    // Prevent adding HTMLPictureElement if it isn't supported (e.g. IE11),
    // but don't prevent it during SSR.
    if (hasPictureElement()) {
      const pic = document.createElement('picture');
      if (selfRef) {
        // Set original component's style.
        pic.width = imageRef.width;
        pic.height = imageRef.height;
      }
      if (hasArtDirectionArray(convertedProps)) {
        const sources = createArtDirectionSources(convertedProps).reverse();
        sources.forEach(currentSource => pic.appendChild(currentSource));
      }

      const sourcesAvif = createSourceElementForSrcSet(imageData, 'avif');
      sourcesAvif && pic.appendChild(sourcesAvif);
      const sourcesWebp = createSourceElementForSrcSet(imageData, 'webp');
      sourcesWebp && pic.appendChild(sourcesWebp);

      pic.appendChild(imageRef);
      bodyClone.appendChild(pic);
    }

    imageRef.sizes = imageData.sizes || ``;
    imageRef.srcset = imageData.srcSet || ``;
    imageRef.src = imageData.src || ``;

    return imageRef;
  }
  return null;
};

/**
 * Creates multiple picture elements.
 *
 * @param imageRefs
 * @param props
 * @param selfRef
 * @return {Array||null}
 */
export const activateMultiplePictureRefs = (imageRefs, props, selfRef) => {
  // Extract Image Array.
  return imageRefs.map((imageRef, index) =>
    activatePictureRef(imageRef, props, selfRef, index, true)
  );
};

/**
 * Checks imageRefs on being active.
 *
 * @param imageRefs
 * @return {*}
 */
export const hasActivatedPictureRefs = imageRefs => {
  return Array.isArray(imageRefs)
    ? imageRefs.every(imageRef => hasPictureRef(imageRef))
    : hasPictureRef(imageRefs);
};

/**
 * Checks imageRef for on being a string or has a currentSrc.
 *
 * @param imageRef
 * @return {boolean}
 */
export const hasPictureRef = imageRef => {
  return isString(imageRef) || (!!imageRef && !!imageRef.currentSrc);
};

/**
 * Checks if an image (array) reference is existing and tests for complete.
 *
 * @param imageRef    HTMLImageElement||array   Image reference(s).
 * @param props
 * @return {boolean}
 */
export const imageReferenceCompleted = (imageRef, props) =>
  imageRef
    ? Array.isArray(imageRef)
      ? imageRef.every(singleImageRef => imageLoaded(singleImageRef)) ||
        inImageCache(props)
      : imageLoaded(imageRef) || inImageCache(props)
    : isString(imageRef);
