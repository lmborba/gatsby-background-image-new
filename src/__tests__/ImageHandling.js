import { switchImageSettings } from '../lib/ImageHandling';
import {
  fixedShapeMock,
  fluidArrayMock,
  fluidShapeMock,
} from './mocks/Various.mock';

describe(`switchImageSettings()`, () => {
  const mockImageRef = {
    src: `http://test.jpg`,
    currentSrc: `http://test.webp`,
    complete: true,
  };
  const state = {
    isVisible: false,
    isLoaded: false,
    imageState: 1,
  };
  it(`should return settings from fluid with empty bgImage`, () => {
    const createdSettings = switchImageSettings({
      image: fluidShapeMock,
      bgImage: ``,
      imageRef: mockImageRef,
      state,
    });
    expect(createdSettings).toMatchSnapshot();
  });

  it(`should return settings from fixed with set bgImage`, () => {
    const createdSettings = switchImageSettings({
      image: fixedShapeMock,
      bgImage: `string_of_base64`,
      imageRef: mockImageRef,
      state,
    });
    expect(createdSettings).toMatchSnapshot();
    expect(createdSettings.lastImage).toEqual(`string_of_base64`);
  });

  it(`should return settings from fixed with set bgImage`, () => {
    const createdSettings = switchImageSettings({
      image: fluidShapeMock,
      bgImage: `string_of_base64`,
      imageRef: mockImageRef,
      state,
    });
    expect(createdSettings).toMatchSnapshot();
  });

  it(`should return settings from fixed array`, () => {
    state.isVisible = true;
    state.imgLoaded = true;
    const createdSettings = switchImageSettings({
      image: fluidArrayMock.fluid,
      bgImage: [`string_of_base64`, `imageString`],
      imageRef: [mockImageRef, mockImageRef],
      state,
    });
    expect(createdSettings).toMatchSnapshot();
  });

  it(`should return settings from fixed array without currentSrc`, () => {
    const mockImageRefNoCurrentSrc = [{ ...mockImageRef }, { ...mockImageRef }];
    delete mockImageRefNoCurrentSrc[0].currentSrc;
    delete mockImageRefNoCurrentSrc[1].currentSrc;
    state.isVisible = true;
    state.imgLoaded = true;
    const createdSettings = switchImageSettings({
      image: fluidArrayMock.fluid,
      imageRef: mockImageRefNoCurrentSrc,
      bgImage: [``, ``],
      state,
    });
    expect(createdSettings).toMatchSnapshot();
  });

  it(`should return settings from fluid without src`, () => {
    const fluid = fluidShapeMock;
    delete fluid.src;
    const mockImageRefNoSrc = { ...mockImageRef };
    delete mockImageRefNoSrc.src;
    state.isVisible = true;
    state.imgLoaded = true;
    const createdSettings = switchImageSettings({
      image: fluid,
      bgImage: ``,
      imageRef: mockImageRefNoSrc,
      state,
    });
    expect(createdSettings).toMatchSnapshot();
  });

  it(`should return settings from fluid with set bgImage without base64`, () => {
    const fluid = fluidShapeMock;
    delete fluid.base64;
    const createdSettings = switchImageSettings({
      image: fluid,
      bgImage: ``,
      imageRef: mockImageRef,
      state,
    });
    expect(createdSettings).toMatchSnapshot();
  });

  it(`should return settings from fluid with set bgImage and tracedSVG`, () => {
    const fluid = {
      ...fluidShapeMock,
      tracedSVG: 'test_tracedSVG.svg',
    };
    delete fluid.base64;
    const createdSettings = switchImageSettings({
      image: fluid,
      bgImage: `test_tracedSVG.svg`,
      imageRef: mockImageRef,
      state,
    });
    expect(createdSettings).toMatchSnapshot();
  });

  it(`should return settings from fluid with set bgImage and base64`, () => {
    const mockRef = mockImageRef;
    delete mockRef.src;
    state.isVisible = true;
    state.isLoaded = true;
    const createdSettings = switchImageSettings({
      image: fluidShapeMock,
      bgImage: `string_of_base64`,
      imageRef: mockRef,
      state,
    });
    expect(createdSettings).toMatchSnapshot();
  });
});
