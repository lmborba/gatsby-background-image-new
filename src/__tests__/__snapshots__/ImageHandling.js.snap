// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`switchImageSettings() should return settings from fixed array 1`] = `
Object {
  "afterOpacity": 1,
  "finalImage": true,
  "lastImage": "string_of_base64,imageString",
  "nextImage": "url('http://test.webp'),url('http://test.webp')",
  "nextImageArray": Array [
    "url('http://test.webp')",
    "url('http://test.webp')",
  ],
}
`;

exports[`switchImageSettings() should return settings from fixed array without currentSrc 1`] = `
Object {
  "afterOpacity": 1,
  "finalImage": true,
  "lastImage": "",
  "nextImage": "url('http://test.jpg'),url('http://test.jpg')",
  "nextImageArray": Array [
    "url('http://test.jpg')",
    "url('http://test.jpg')",
  ],
}
`;

exports[`switchImageSettings() should return settings from fixed with set bgImage 1`] = `
Object {
  "afterOpacity": 1,
  "finalImage": false,
  "lastImage": "string_of_base64",
  "nextImage": "url(string_of_base64)",
}
`;

exports[`switchImageSettings() should return settings from fixed with set bgImage 2`] = `
Object {
  "afterOpacity": 1,
  "finalImage": false,
  "lastImage": "string_of_base64",
  "nextImage": "url(string_of_base64)",
}
`;

exports[`switchImageSettings() should return settings from fluid with empty bgImage 1`] = `
Object {
  "afterOpacity": 1,
  "finalImage": false,
  "lastImage": "",
  "nextImage": "url(string_of_base64)",
}
`;

exports[`switchImageSettings() should return settings from fluid with set bgImage and base64 1`] = `
Object {
  "afterOpacity": 1,
  "finalImage": true,
  "lastImage": "string_of_base64",
  "nextImage": "url('http://test.webp')",
}
`;

exports[`switchImageSettings() should return settings from fluid with set bgImage and tracedSVG 1`] = `
Object {
  "afterOpacity": 1,
  "finalImage": true,
  "lastImage": "test_tracedSVG.svg",
  "nextImage": "url('http://test.webp')",
}
`;

exports[`switchImageSettings() should return settings from fluid with set bgImage without base64 1`] = `
Object {
  "afterOpacity": 1,
  "finalImage": true,
  "lastImage": "",
  "nextImage": "url('http://test.webp')",
}
`;

exports[`switchImageSettings() should return settings from fluid without src 1`] = `
Object {
  "afterOpacity": 1,
  "finalImage": true,
  "lastImage": "",
  "nextImage": "url('http://test.webp')",
}
`;
